### DISCUSSION TITLE
FakePlayer issue

### DISCUSSION THREAD

**Discussion by para-dise** [Minestom/Minestom#830] [Help]
So,

I have the following code: 
```java
FakePlayerOption option = new FakePlayerOption();
option.setInTabList(false);
option.setRegistered(false);
FakePlayer.initPlayer(
    UUID.fromString("user1_uuid"),
    "user1",
    option,
    (FakePlayer p) -> {
        p.teleport(new Pos(-12.445, 67.0, 118.28));
    }
);
``` 
which throws 
```
java.lang.IllegalStateException: You need to use Entity#setInstance before teleporting an entity!
at net.minestom.server.utils.validate.Check.stateCondition(Check.java:55)
at net.minestom.server.entity.Entity.teleport(Entity.java:295)
at net.minestom.server.entity.Entity.teleport(Entity.java:319)
```

so then I tried to set an instance with 

```java
FakePlayerOption option = new FakePlayerOption();
option.setInTabList(false);
option.setRegistered(false);
FakePlayer.initPlayer(
    UUID.fromString("user1_uuid"),
    "user1",
    option,
    (FakePlayer p) -> {
        p.setInstance(this.instance);
        p.teleport(new Pos(-12.445, 67.0, 118.28));
    }
);
```

which then gives me the error `java.lang.IllegalArgumentException: Instance should be different than the current one`, so I mean, am I doing something wrong or should I slap this in a try/catch and ignore it..? The FakePlayers are successfully spawning, but the error is pretty annoying

**Comment by GreatWyrm** [comment]
Seems rather strange, you might be setting their instance twice somewhere, I would maybe add some debug statements before your cases of setInstance so you know how many times the section of code is getting called

I would also recommend using the `Entity.setInstance(Instance, Pos)` function instead of doing them separately, so your code would look like this:

```
(FakePlayer p) -> {
        p.setInstance(this.instance, new Pos(-12.445, 67.0, 118.28));
 }
 ```
 
 Lastly, I would not recommend try/catching that error.

**Reply by para-dise** [reply]
Well, I tried 
```java
(FakePlayer p) -> {
    System.out.println("setting instance");
    p.setInstance(this.instance, new Pos(-9.5, 67.0, 123.5));
    p.lookAt(this.lookAt);
}
```

And I see only one occurrence of `setting instance` in console and then it throws `java.lang.IllegalArgumentException: Instance should be different than the current one` again

**Comment by Eoghanmc22** [comment]
Be careful with fake player. Fake player is more similar to an actual tcp connection than a npc (it doesnt actually matain a network connection but it is treated as if its a real player, triggers all events a player would, and can communicate with packets) you most likely want to create an entity with the entity type of player
